File: counting-bloom-filter/counting-bloom-filter.cc
# include <bitset>

const int MAXN = 200000;
const int MOD = 100000;

class HashFunction {
    public:
    HashFunction(int m, int b): mod(m), base(b) {}
    int get_hash(const string &s) {
        int ans = 0;
        for (int i = 0; s[i]; ++i) {
            ans = (ans * base + s[i]) % mod;
        }
        return ans;
    }
    
    private:
    int mod, base;
};

class CountingBloomFilter {
    public:
    CountingBloomFilter(int k) {
        // initialize your data structure here
        int mod = MOD, base = 3;
        for (int i = 0; i < k; ++i) {
            HashFunction *hf = new HashFunction(mod, base);
            hfs.push_back(hf);
            ++mod;
            base += 2;
        }
        n_hash = k;
        cnts = vector<int>(MAXN, 0);
    }
    
    void add(string& word) {
        // Write your code here
        for (int i = 0; i < n_hash; ++i) {
            int h = hfs[i]->get_hash(word);
            ++cnts[h];
        }
    }
    
    void remove(string& word) {
        // Write your code here
        for (int i = 0; i < n_hash; ++i) {
            int h = hfs[i]->get_hash(word);
            --cnts[h];
        }
    }
    bool contains(string& word) {
        // Write your code here
        for (int i = 0; i < n_hash; ++i) {
            int h = hfs[i]->get_hash(word);
            if (!cnts[h]) return false;
        }
        return true;
    }
    
    private:
    int n_hash;
    vector<HashFunction*> hfs;
    vector<int> cnts;
};

File: counting-bloom-filter/counting-bloom-filter.py
class HashFunction(object):
    def __init__(self, m, b):
        self.mod = m
        self.base = b
    
    def get_hash(self, s):
        ans = 0
        for i in xrange(len(s)):
            ans = (ans * self.base + ord(s[i])) % self.mod
        return ans

class CountingBloomFilter:
    
    # @param {int} k an integer
    def __init__(self, k):
        # initialize your data structure here
        self.k = k
        self.hfs = []
        for i in xrange(k):
            self.hfs.append(HashFunction(100000 + i, 2 * i + 3))
        self.bits =  [0] * 200000
    
    
    # @param {str} word a string
    def add(self, word):
        # Write your code here
        for i in xrange(self.k):
            h = self.hfs[i].get_hash(word)
            self.bits[h] += 1

# @param {str} word a string
def remove(self, word):
    # Write your code here
    for i in xrange(self.k):
        h = self.hfs[i].get_hash(word)
            self.bits[h] -= 1

# @param {str} word a string
# @return {bool} True if word is exists in bllom filter or false
def contains(self, word):
    # Write your code here
    for i in xrange(self.k):
        h = self.hfs[i].get_hash(word)
            if not self.bits[h]:
                return False
    return True
